### ISSUE TITLE
Ice spikes do not generate in Ice Plains Spikes biomes

### ISSUE THREAD

**Issue by hoqhuuep** [SpongePowered/SpongeForge#548]
There also seems to be an issue with the spruce trees in the background, these are in either Ice Plains or Ice Mountains biome.

Exact version of SpongeForge tested: spongeforge-1.8.9-1732-3.1.0-BETA-1139.jar

With SpongeForge installed:
![2016-02-22_10 08 45](https://cloud.githubusercontent.com/assets/2024532/13206192/27a866ca-d94d-11e5-9c02-4a2d7fc4f1ae.png)

Without SpongeForge installed (same version of MinecraftForge, just removed SpongeForge from mods folder):
![2016-02-22_10 11 48](https://cloud.githubusercontent.com/assets/2024532/13206193/28bd7118-d94d-11e5-9502-70fe216655d9.png)

World seed: -479955565
Coordinates: See screenshots

**Comment by hoqhuuep** [comment]
If someone is able to point me to the code responsible for setting the biome specific populators in both vanilla and SpongeForge, I'd be happy to do a review to check for any other discrepancies.

**Comment by gabizou** [comment]
Your SF is a bit out of date, could you re-test on the latest (which is 1159)?

**Comment by hoqhuuep** [comment]
Wow, that was a fast response! Yep. Will post back in a few minutes!

**Comment by Deamon5550** [comment]
[Ice Spikes are added here](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/mixin/core/world/biome/MixinBiomeGenSnow.java#L48) so some testing needed obviously to see why they aren't working.

**Comment by hoqhuuep** [comment]
Thanks for the link to the code, I'll have a dig around and see if I can spot anything.

Tested again with: spongeforge-1.8.9-1732-3.1.0-BETA-1159.jar

Same result.
![2016-02-22_10 29 50](https://cloud.githubusercontent.com/assets/2024532/13206311/62cf9e88-d94f-11e5-8c8a-9edaa35301ff.png)

**Comment by hoqhuuep** [comment]
I can confirm that this is fixed in: spongeforge-1.8.9-1732-3.1.0-BETA-1165.jar
Both the ice spikes and the spruce trees are generating. Nice work @Minecrell !

I notice that the exact positions of the spikes and trees do not match the generation without Sponge installed. I assume this is to do with the order that the populators are invoked or how the random number generator is used. I am unsure if Sponge has a design goal to produce exactly the same terrain or just close enough, so I don't know if this is a bug.

Compare this image (using the fixed spongeforge-1.8.9-1732-3.1.0-BETA-1165.jar) with the above image (without Sponge installed):
![2016-02-25_18 37 58](https://cloud.githubusercontent.com/assets/2024532/13313140/d93d011c-dbef-11e5-8fbb-7c5e8749b3a8.png)

**Comment by Zidane** [comment]
@Deamon5550 

Please investigate the differences in generation.

**Comment by ryantheleach** [comment]
@hoqhuuep Have you created a control test of vanilla / vanilla / forge? Do ice spikes naturally vary given the same seed on vanilla? could this be a forge bug?

**Comment by ryantheleach** [comment]
@hoqhuuep You have tested spongeforge against forge, but have you tested forge vs forge or vanilla vs forge? It may be that ice spikes naturally change and don't perfectly replicate between worlds with the same seed?

**Comment by hoqhuuep** [comment]
I have tested Forge vs Forge, generation is definitely the same each time. I will check other combinations and post back. If the intention is that it should be identical to vanilla, I can raise this as a separate bug (?)

**Comment by hoqhuuep** [comment]
While I'm at it, I'll check the behavior of some other populators, e.g. cactus, trees in other biomes, etc.

**Comment by gabizou** [comment]
Populator object placement is random each and every time, regardless of whether the seed is the same. This is due to the use of `java.util.Random` and variance of object chances being placed. So yes, the variance in where the spikes are to be placed is to be expected (even in vanilla cases and forge should respect it as well).

**Comment by Deamon5550** [comment]
things like terrain and biomes will be identical, but the exact placement of things like trees and such may be slightly different, this is because of a few changes to the populators and the number of calls to the random object. This is entire expected and is really a non-issue.

**Comment by hoqhuuep** [comment]
I was about to disagree with you. But on closer inspection I believe you are correct.

Here is an image of the same scene in vanilla. At first glance it looks the same as the one I created with Forge (and it is certainly closer than the one created with Sponge). However with closer inspection it is clear that there are small differences (circled some in red).

![2016-02-25_19 25 09-2](https://cloud.githubusercontent.com/assets/2024532/13313938/ef652374-dbf5-11e5-8214-e48a3576a928.png)

I repeated this test in vanilla a second time, and the placement of the ice spikes was identical to the above image. However there were some differences in the placement of tall grass in the background.

Therefore, I'm quite sure the intention is to create an identical world every time, but in some cases the random generators are not seeded quite right, even in vanilla.
